File: Management/Nation.js
// File = robjam1990/Psychosis/Gameplay/Management/Nation.js
class Nation {
    constructor(name, population, militaryStrength, industry, agriculture, commerce) {
        this.name = name;
        this.population = population;
        this.militaryStrength = militaryStrength;
        this.industry = industry;
        this.agriculture = agriculture;
        this.commerce = commerce;
        this.territories = [];
        this.currentLeader = null;
        this.time = {
            hours: 0,
            day: 0,
            season: "Spring" // Initialize season as Spring
        };
        this.taxRate = 0.1; // Initial tax rate
        this.tradePartners = []; // Array to store trade partners
        this.diplomaticRelations = {}; // Object to store diplomatic relations with other nations
        this.culturalDevelopment = 0; // Cultural development level
        this.socialPolicies = {}; // Object to store enacted social policies
        this.eventDescriptions = [
            "A natural disaster strikes!",
            "A diplomatic crisis erupts with a neighboring nation!",
            "A new technological breakthrough boosts industry!",
            "A cultural renaissance inspires the population!",
            "A plague sweeps through the land, affecting agriculture and population!",
            "A rebellion threatens the stability of the nation!",
            "A bountiful harvest leads to economic prosperity!",
            "A military victory boosts national morale!"
        ]; // Descriptions of random events
    }

    // Method to add a territory to the nation
    addTerritory(territory) {
        this.territories.push(territory);
    }

    // Method to remove a territory from the nation
    removeTerritory(territory) {
        const index = this.territories.indexOf(territory);
        if (index !== -1) {
            this.territories.splice(index, 1);
        }
    }

    // Method to manage logistics
    manageLogistics() {
        // Placeholder: Implement logistics management logic
        console.log("Managing logistics...");
        // Implement transportation routes, supply chains, infrastructure development, etc.
    }

    // Method to manage agriculture
    manageAgriculture() {
        // Placeholder: Implement agriculture management logic
        console.log("Managing agriculture...");
        // Implement crop rotation, irrigation systems, etc.
    }

    // Method to manage industry
    manageIndustry() {
        // Placeholder: Implement industry management logic
        console.log("Managing industry...");
        // Implement mining, manufacturing, technological development, etc.
    }

    // Method to manage taxation
    manageTaxation(newTaxRate) {
        // Placeholder: Implement taxation management logic
        console.log(`Adjusting tax rate to ${newTaxRate}...`);
        this.taxRate = newTaxRate;
    }

    // Method to manage trade agreements
    manageTradeAgreements(partnerNation) {
        // Placeholder: Implement trade agreement management logic
        console.log(`Establishing trade agreement with ${partnerNation.name}...`);
        this.tradePartners.push(partnerNation);
    }

    // Method to manage diplomacy
    manageDiplomacy(action, targetNation) {
        // Placeholder: Implement diplomacy management logic
        console.log(`Initiating ${action} with ${targetNation.name}...`);
        // Implement diplomatic actions such as alliance, ceasefire, declaration of war, etc.
    }

    // Method to enact social policies
    enactSocialPolicy(policy, value) {
        // Placeholder: Implement social policy management logic
        console.log(`Enacting ${policy} policy with value ${value}...`);
        this.socialPolicies[policy] = value;
    }

    // Method to handle random events
    handleEvents() {
        // Placeholder: Implement event handling logic
        const eventIndex = Math.floor(Math.random() * this.eventDescriptions.length);
        const eventDescription = this.eventDescriptions[eventIndex];
        console.log(eventDescription);
        // Implement event consequences
        switch (eventIndex) {
            case 0: // Natural disaster
                this.population *= 0.9; // 10% population decrease
                this.militaryStrength *= 0.8; // 20% military strength decrease
                break;
            case 2: // Technological breakthrough
                this.industry += 1000; // Boost industry by 1000 units
                break;
            case 3: // Cultural renaissance
                this.culturalDevelopment += 10; // Increase cultural development level
                break;
            case 4: // Plague
                this.population *= 0.7; // 30% population decrease
                this.agriculture *= 0.8; // 20% agriculture decrease
                break;
            case 5: // Rebellion
                this.militaryStrength *= 0.7; // 30% military strength decrease
                break;
            case 6: // Bountiful harvest
                this.agriculture += 2000; // Boost agriculture by 2000 units
                break;
            case 7: // Military victory
                this.militaryStrength *= 1.2; // 20% military strength increase
                break;
            default:
                break;
        }
    }

    // Method to advance time
    advanceTime(gameHours) {
        const daysPassed = Math.floor(gameHours / 24);

        for (let i = 0; i < daysPassed; i++) {
            this.manageAgriculture();
            this.manageIndustry();
            this.handleEvents(); // Handle events each day
            this.updatePopulation();
            this.updateMilitaryStrength();
            this.updateTime(); // Update time including day, season, etc.
        }
    }

    // Method to update time including day, season, etc.
    updateTime() {
        this.time.hours += 24;
        this.time.day++;
        if (this.time.day % 30 === 0) { // Check if a month has passed
            this.time.day = 1;
            this.updateSeason();
        }
    }

    // Method to update the season
    updateSeason() {
        const seasons = ["Spring", "Summer", "Autumn", "Winter"];
        const currentSeasonIndex = seasons.indexOf(this.time.season);
        const nextSeasonIndex = (currentSeasonIndex + 1) % 4;
        this.time.season = seasons[nextSeasonIndex];
    }
}

// Example usage:

// Initialize the nation and leader
const playerNation = new Nation("Bractalia", { citizens: 10000, farmers: 2000, miners: 1500 }, 1000, 5000, 3000, 2000);
const playerLeader = new Leader("Apus", "Queen", 3);
playerNation.currentLeader = playerLeader;

// Add territories to the nation
playerNation.addTerritory("Territory 1");
playerNation.addTerritory("Territory 2");

// Game loop
let isGameOver = false;
while (!isGameOver) {
    // Display menu options and gather player input
    console.log("1. Manage Logistics");
    console.log("2. Manage Agriculture");
    console.log("3. Manage Commerce");
    console.log("4. Manage Military");
    console.log("5. Manage Taxation");
    console.log("6. Manage Trade Agreements");
    console.log("7. Manage Diplomacy");
    console.log("8. Enact Social Policies");
    console.log("9. End Turn");
    console.log("10. Exit Game");
    const choice = parseInt(prompt("Enter your choice: "));

    // Execute player's choice
    switch (choice) {
        case 1:
            playerNation.manageLogistics();
            break;
        case 2:
            playerNation.manageAgriculture();
            break;
        case 3:
            playerNation.manageCommerce();
            break;
        case 4:
            playerNation.manageMilitary();
            break;
        case 5:
            const newTaxRate = parseFloat(prompt("Enter the new tax rate: "));
            playerNation.manageTaxation(newTaxRate);
            break;
        case 6:
            const partnerNation = prompt("Enter the name of the nation to establish a trade agreement with: ");
            playerNation.manageTradeAgreements(partnerNation);
            break;
        case 7:
            const action = prompt("Enter the diplomatic action (e.g., alliance, ceasefire, declaration of war): ");
            const targetNation = prompt("Enter the name of the target nation: ");
            playerNation.manageDiplomacy(action, targetNation);
            break;
        case 8:
            const policy = prompt("Enter the social policy to enact: ");
            const value = prompt("Enter the value of the social policy: ");
            playerNation.enactSocialPolicy(policy, value);
            break;
        case 9:
            console.log("Ending turn...");
            // Handle events and advance time by one day (24 game hours)
            playerNation.handleEvents();
            playerNation.advanceTime(24);
            break;
        case 10:
            console.log("Exiting game...");
            isGameOver = true; // Set the flag to end the game loop
            break;
        default:
            console.log("Invalid choice.");
            break;
    }
}

File: Management/Nation.py
import random

class Nation:
    def __init__(self, name, population, military_strength, industry, agriculture, commerce):
        self.name = name
        self.population = population
        self.military_strength = military_strength
        self.industry = industry
        self.agriculture = agriculture
        self.commerce = commerce
        self.territories = []
        self.current_leader = None
        self.time = {
            "hours": 0,
            "day": 0,
            "season": "Spring"  # Initialize season as Spring
        }
        self.tax_rate = 0.1  # Initial tax rate
        self.trade_partners = []  # Array to store trade partners
        self.diplomatic_relations = {}  # Dict to store diplomatic relations with other nations
        self.cultural_development = 0  # Cultural development level
        self.social_policies = {}  # Dict to store enacted social policies
        self.event_descriptions = [
            "A natural disaster strikes!",
            "A diplomatic crisis erupts with a neighboring nation!",
            "A new technological breakthrough boosts industry!",
            "A cultural renaissance inspires the population!",
            "A plague sweeps through the land, affecting agriculture and population!",
            "A rebellion threatens the stability of the nation!",
            "A bountiful harvest leads to economic prosperity!",
            "A military victory boosts national morale!"
        ]  # Descriptions of random events

    # Method to add a territory to the nation
    def add_territory(self, territory):
        self.territories.append(territory)

    # Method to remove a territory from the nation
    def remove_territory(self, territory):
        if territory in self.territories:
            self.territories.remove(territory)

    # Method to manage logistics
    def manage_logistics(self):
        # Placeholder: Implement logistics management logic
        print("Managing logistics...")
        # Implement transportation routes, supply chains, infrastructure development, etc.

    # Method to manage agriculture
    def manage_agriculture(self):
        # Placeholder: Implement agriculture management logic
        print("Managing agriculture...")
        # Implement crop rotation, irrigation systems, etc.

    # Method to manage industry
    def manage_industry(self):
        # Placeholder: Implement industry management logic
        print("Managing industry...")
        # Implement mining, manufacturing, technological development, etc.

    # Method to manage taxation
    def manage_taxation(self, new_tax_rate):
        # Placeholder: Implement taxation management logic
        print(f"Adjusting tax rate to {new_tax_rate}...")
        self.tax_rate = new_tax_rate

    # Method to manage trade agreements
    def manage_trade_agreements(self, partner_nation):
        # Placeholder: Implement trade agreement management logic
        print(f"Establishing trade agreement with {partner_nation.name}...")
        self.trade_partners.append(partner_nation)

    # Method to manage diplomacy
    def manage_diplomacy(self, action, target_nation):
        # Placeholder: Implement diplomacy management logic
        print(f"Initiating {action} with {target_nation.name}...")
        # Implement diplomatic actions such as alliance, ceasefire, declaration of war, etc.

    # Method to enact social policies
    def enact_social_policy(self, policy, value):
        # Placeholder: Implement social policy management logic
        print(f"Enacting {policy} policy with value {value}...")
        self.social_policies[policy] = value

    # Method to handle random events
    def handle_events(self):
        # Placeholder: Implement event handling logic
        event_index = random.randint(0, len(self.event_descriptions) - 1)
        event_description = self.event_descriptions[event_index]
        print(event_description)
        # Implement event consequences
        if event_index == 0:  # Natural disaster
            self.population *= 0.9  # 10% population decrease
            self.military_strength *= 0.8  # 20% military strength decrease
        elif event_index == 2:  # Technological breakthrough
            self.industry += 1000  # Boost industry by 1000 units
        elif event_index == 3:  # Cultural renaissance
            self.cultural_development += 10  # Increase cultural development level
        elif event_index == 4:  # Plague
            self.population *= 0.7  # 30% population decrease
            self.agriculture *= 0.8  # 20% agriculture decrease
        elif event_index == 5:  # Rebellion
            self.military_strength *= 0.7  # 30% military strength decrease
        elif event_index == 6:  # Bountiful harvest
            self.agriculture += 2000  # Boost agriculture by 2000 units
        elif event_index == 7:  # Military victory
            self.military_strength *= 1.2  # 20% military strength increase

    # Method to advance time
    def advance_time(self, game_hours):
        days_passed = game_hours // 24

        for _ in range(days_passed):
            self.manage_agriculture()
            self.manage_industry()
            self.handle_events()  # Handle events each day
            self.update_time()  # Update time including day, season, etc.

    # Method to update time including day, season, etc.
    def update_time(self):
        self.time["hours"] += 24
        self.time["day"] += 1
        if self.time["day"] % 30 == 0:  # Check if a month has passed
            self.time["day"] = 1
            self.update_season()

    # Method to update the season
    def update_season(self):
        seasons = ["Spring", "Summer", "Autumn", "Winter"]
        current_season_index = seasons.index(self.time["season"])
        next_season_index = (current_season_index + 1) % 4
        self.time["season"] = seasons[next_season_index]

# Example usage:

# Initialize the nation and leader
player_nation = Nation("Bractalia", {"citizens": 10000, "farmers": 2000, "miners": 1500}, 1000, 5000, 3000, 2000)
player_leader = Leader("Apus", "Queen", 3)
player_nation.current_leader = player_leader

# Add territories to the nation
player_nation.add_territory("Territory 1")
player_nation.add_territory("Territory 2")

# Game loop
is_game_over = False
while not is_game_over:
    # Display menu options and gather player input
    print("1. Manage Logistics")
    print("2. Manage Agriculture")
    print("3. Manage Commerce")
    print("4. Manage Military")
    print("5. Manage Taxation")
    print("6. Manage Trade Agreements")
    print("7. Manage Diplomacy")
    print("8. Enact Social Policies")
    print("9. End Turn")
    print("10. Exit Game")
    choice = int(input("Enter your choice: "))

    # Execute player's choice
    if choice == 1:
        player_nation.manage_logistics()
    elif choice == 2:
        player_nation.manage_agriculture()
    elif choice == 3:
        player_nation.manage_commerce()
    elif choice == 4:
        player_nation.manage_military()
    elif choice == 5:
        new_tax_rate = float(input("Enter the new tax rate: "))
        player_nation.manage_taxation(new_tax_rate)
    elif choice == 6:
        partner_nation = input("Enter the name of the nation to establish a trade agreement with: ")
        player_nation.manage_trade_agreements(partner_nation)
    elif choice == 7:
        action = input("Enter the diplomatic action (e.g., alliance, ceasefire, declaration of war): ")
        target_nation = input("Enter the name of the target nation: ")
        player_nation.manage_diplomacy(action, target_nation)
    elif choice == 8:
        policy = input("Enter the social policy to enact: ")
        value = input("Enter the value of the social policy: ")
        player_nation.enact_social_policy(policy, value)
    elif choice == 9:
        print("Ending turn...")
        # Handle events and advance time by one day (24 game hours)
        player_nation.handle_events()
        player_nation.advance_time(24)
    elif choice == 10:
        print("Exiting game...")
        is_game_over = True  # Set the flag to end the game loop
    else:
        print("Invalid choice.")
